Userland/Services/NotificationServer/ClientConnection.cpp

```diff
@@ -57,7 +57,41 @@ OwnPtr<Messages::NotificationServer::ShowNotificationResponse> ClientConnection:
 {
     auto window = NotificationWindow::construct(message.text(), message.title(), message.icon());
     window->show();
-    return make<Messages::NotificationServer::ShowNotificationResponse>();
+    return make<Messages::NotificationServer::ShowNotificationResponse>(window->get_window_id());
+}
+
+OwnPtr<Messages::NotificationServer::UpdateNotificationIconResponse> ClientConnection::handle(const Messages::NotificationServer::UpdateNotificationIcon& message)
+{
+    auto window = NotificationWindow::get_window_by_id(message.id());
+    if (window) {
+        window->set_image(message.icon());
+    }
+    return make<Messages::NotificationServer::UpdateNotificationIconResponse>(window);
+}
+
+OwnPtr<Messages::NotificationServer::UpdateNotificationTextResponse> ClientConnection::handle(const Messages::NotificationServer::UpdateNotificationText& message)
+{
+    auto window = NotificationWindow::get_window_by_id(message.id());
+    if (window) {
+        window->set_text(message.text());
+        window->set_title_widget(message.title());
+    }
+    return make<Messages::NotificationServer::UpdateNotificationTextResponse>(window);
+}
+
+OwnPtr<Messages::NotificationServer::RemoveNotificationResponse> ClientConnection::handle(const Messages::NotificationServer::RemoveNotification& message)
+{
+    auto window = NotificationWindow::get_window_by_id(message.id());
+    if (window) {
+        window->close();
```

Comment:
I think this leaks memory by not removing the window from `s_windows`:
- `window->close()` is only implemented by `GUI::Window::close`
- `GUI::Window::on_close` is never set, and does not exist by default (and will not call `on_close_request`)

Is that the case? If so, maybe changing `NotificationWindow` to hook `on_close` instead of `on_close_request` makes sense?